### PR TITLE
Directly generate batch without intermediate entries

### LINKED ISSUE #10: Generate `Batch` without intermediate steps
ATM, we generate a large slice of `Entry` and then convert them into a `Batch`.
To optimize allocations and CPU usage, generating the step could be done directly.

### PR DESCRIPTION
**Pull request by MStoykov**
Fixes #10

### CODE REVIEW

**Review by vlad-diachenko — APPROVED** [review]
LGTM

File: batch.go [hunk 1]
```diff
@@ -122,41 +121,37 @@ func (b *Batch) createPushRequest() (*logproto.PushRequest, int) {
...
-			streamSize += len(line)
-			lines++
+		stream := &logproto.Stream{Labels: labels.String()}
+		batch.Streams[stream.Labels] = stream
+
+		var line string
+		for ; batch.Bytes < maxSizePerStream; batch.Bytes += len(line) {
+			line = flog.NewLog(string(labels[model.LabelName("format")]), time.Now())
+			stream.Entries = append(stream.Entries, logproto.Entry{
+				Timestamp: time.Now(),
```
Comment (MStoykov) [inline batch.go:150]: These two calls to time.Now() are currently the *biggest* CPU user if you just run the `examples/simple.js` - we can make it into one in the begging of the `for` loop, but I guess you want to have the time in the line?

File: batch.go [hunk 2]
```diff
@@ -122,41 +121,38 @@ func (b *Batch) createPushRequest() (*logproto.PushRequest, int) {
...
+		batch.Streams[stream.Labels] = stream
+
+		var line string
+		for ; batch.Bytes < maxSizePerStream; batch.Bytes += len(line) {
+			now := time.Now()
+			line = flog.NewLog(string(labels[model.LabelName("format")]), now)
+			stream.Entries = append(stream.Entries, logproto.Entry{
+				Timestamp: now,
+				Line:      line,
+			})
```
Comment (chaudum) [inline batch.go:153]:
```suggestion
        var now time.Timestamp
		logFmt := string(labels[model.LabelName("format")])
		for ; batch.Bytes < maxSizePerStream; batch.Bytes += len(line) {
			now = time.Now()
			stream.Entries = append(stream.Entries, logproto.Entry{
				Timestamp: now,
				Line:      flog.NewLog(logFmt, now),
			})
```

**Comment by chaudum** [comment]
In general LGTM!

**Comment by MStoykov** [comment]
I did add a basic benchmark, I would've made it a table-driven one and added a bunch of variants, but I have no idea what will be good values for that, so if you provide more I will update it.

I can rearrange and squash commits if we want to keep some kind of history or not if we are just going to merge squash this in the end - I don't know what you prefer, but I have no problems with either.

**Comment by chaudum** [comment]
> I can rearrange and squash commits if we want to keep some kind of history or not if we are just going to merge squash this in the end - I don't know what you prefer, but I have no problems with either.

We can merge squash the commits.

> p.s. the commit message has the results

They look great!

Great work! :partying_face:

**Review by chaudum — APPROVED** [review]

**Comment by MStoykov** [comment]
@chaudum merge when you want if you don't want more changes

**Pull request merged**